`.env` Hot Reloading Error (Astro v5 Beta)

### Astro Info

```block
Astro                    v5.0.0-beta.3
Node                     v20.17.0
System                   macOS (arm64)
Package Manager          pnpm
Output                   static
Adapter                  none
Integrations             none
```


### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

When modifying a `.env` file while the dev server is running, a Vite error is logged to the console.

```
[ERROR] [vite] Error: cannot print server URLs before server.listen is called.
```

Sometimes (I can't tell if it's when you make multiple saves in rapid succession or something else), you also get the following Node error thrown.

```
(node:84966) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 unhandledRejection listeners added to [process]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
```

As far as I can tell (and I'm still using environment variables the old way) the environment variables, both with the `PUBLIC_` prefix and without, still work.

### What's the expected result?

I don't think these warnings, if true, are supposed to print.

### Link to Minimal Reproducible Example

https://github.com/neelsudhakaran/astro-server-listen-error

### Participation

- [X] I am willing to submit a pull request for this issue.

--- a/.changeset/dull-moles-talk.md
+++ b/.changeset/dull-moles-talk.md
@@ -1,0 +1,5 @@
+---
+'astro': patch
+---
+
+Prevents Vite emitting an error when restarting itself

--- a/packages/astro/src/core/dev/restart.ts
+++ b/packages/astro/src/core/dev/restart.ts
@@ -136,7 +136,7 @@ export async function createContainerWithAutomaticRestart({
 		},
 	};
 
-	async function handleServerRestart(logMsg = '') {
+	async function handleServerRestart(logMsg = '', server?: vite.ViteDevServer) {
 		logger.info(null, (logMsg + ' Restarting...').trim());
 		const container = restart.container;
 		const result = await restartContainer(container);
@@ -147,6 +147,11 @@ export async function createContainerWithAutomaticRestart({
 			// Restart success. Add new watches because this is a new container with a new Vite server
 			restart.container = result;
 			setupContainer();
+			if (server) {
+				// Vite expects the resolved URLs to be available
+				server.resolvedUrls = result.viteServer.resolvedUrls;
+			}
+
 			resolveRestart(null);
 		}
 		restartComplete = new Promise<Error | null>((resolve) => {
@@ -171,7 +176,8 @@ export async function createContainerWithAutomaticRestart({
 
 		// Restart the Astro dev server instead of Vite's when the API is called by plugins.
 		// Ignore the `forceOptimize` parameter for now.
-		restart.container.viteServer.restart = () => handleServerRestart();
+		restart.container.viteServer.restart = () =>
+			handleServerRestart('', restart.container.viteServer);
 
 		// Set up shortcuts
 

